go/vt/vtgate/engine/send.go

```diff
@@ -110,12 +110,39 @@ func (s *Send) Execute(vcursor VCursor, bindVars map[string]*querypb.BindVariabl
 }
 
 // StreamExecute implements Primitive interface
-func (s *Send) StreamExecute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantields bool, callback func(*sqltypes.Result) error) error {
-	return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "not reachable") // TODO: systay - this should work
+func (s *Send) StreamExecute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
+	rss, _, err := vcursor.ResolveDestinations(s.Keyspace.Name, nil, []key.Destination{s.TargetDestination})
+	if err != nil {
+		return vterrors.Wrap(err, "sendStreamExecute")
+	}
+
+	if !s.Keyspace.Sharded && len(rss) != 1 {
+		return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "Keyspace does not have exactly one shard: %v", rss)
+	}
+
```

Comment:
Post this check. Add check for

```
if s.SingleShardOnly && len(rss) != 1 {
	return nil, 
	       vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, 
	                       "Unexpected error, DestinationKeyspaceID mapping to multiple shards: %s, got: %v", 
	                       s.Query, 
	                       s.TargetDestination)
}

```
There are query plans which should be sent to only single shard.

Reply:
Updated, plus unit tests for this check for OLTP and OLAP